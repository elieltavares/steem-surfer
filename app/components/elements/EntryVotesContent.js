import React, { Component } from 'react';

import PropTypes from 'prop-types';

import { FormattedMessage } from 'react-intl';

import FormattedCurrency from './FormattedCurrency';

class EntryVotesContent extends Component {
  render() {
    const { votes } = this.props;

    const popoverVotes = votes.slice(0, 10);
    const moreCount = votes.length - 10;

    const popoverChildren = popoverVotes.map(v => (
      <p key={v.voter}>
        <span key="voter" className="voter">
          {v.voter}
        </span>
        <span key="percent" className="percent">
          {v.percent.toFixed(1)}
          {'%'}
        </span>
        <span key="reward" className="reward">
          <FormattedCurrency {...this.props} value={v.reward} fixAt={3} />
        </span>
      </p>
    ));

    if (moreCount > 0) {
      popoverChildren.push(
        <p key="more" className="more">
          <FormattedMessage id="entry-votes.n-more" values={{ n: moreCount }} />
        </p>
      );
    }

    return <div className="votes-popover-content">{popoverChildren}</div>;
  }
}

EntryVotesContent.propTypes = {
  votes: PropTypes.arrayOf(PropTypes.object).isRequired
};

export default EntryVotesContent;
