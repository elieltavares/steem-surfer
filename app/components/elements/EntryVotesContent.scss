@import '../../style/mixins';
@import '../../style/colors';

.votes-popover-content {
  width: 320px;

  p {
    margin: 0 0 12px 0;
    display: flex;

    &:last-of-type {
      margin-bottom: 0;
    }

    .voter {
      font-weight: 700;
      margin-right: 8px;
    }

    .percent {
      @include themify(day) {
        color: $warm-grey;
      }

      @include themify(night) {
        color: $steel-grey;
      }
    }

    .reward {
      flex-grow: 1;
      text-align: right;
    }
  }

  .more {
    font-weight: 500;
  }
}
