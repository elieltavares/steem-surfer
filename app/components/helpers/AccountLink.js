import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { getAccount } from '../../backend/steem-client';

export const makePath = username => `/@${username}`;

class AccountLink extends Component {
  goProfile = async () => {
    const { username, history, actions, onClick, afterClick } = this.props;

    onClick();

    let { accountData } = this.props;
    const { setVisitingAccount } = actions;

    if (!accountData) {
      try {
        accountData = await getAccount(username);
      } catch (err) {
        accountData = null;
      }
    }

    if (accountData) {
      setVisitingAccount(accountData);
    }

    history.push(makePath(username));

    afterClick();
  };

  render() {
    const { children } = this.props;

    return React.cloneElement(children, {
      onClick: this.goProfile
    });
  }
}

AccountLink.defaultProps = {
  accountData: null,
  onClick: () => {},
  afterClick: () => {}
};

AccountLink.propTypes = {
  children: PropTypes.element.isRequired,
  username: PropTypes.string.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  accountData: PropTypes.instanceOf(Object),
  onClick: PropTypes.func,
  afterClick: PropTypes.func,
  actions: PropTypes.shape({
    setVisitingAccount: PropTypes.func.isRequired
  }).isRequired
};

export default AccountLink;
