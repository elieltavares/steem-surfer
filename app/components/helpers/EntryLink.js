import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { getContent } from '../../backend/steem-client';

export const makePath = (category, author, permlink, toReplies = false) =>
  `/${category}/@${author}/${permlink}${toReplies ? '#replies' : ''}`;

class EntryLink extends Component {
  goEntry = async () => {
    let { entry } = this.props;
    const { history, actions, toReplies, markAsRead } = this.props;

    if (markAsRead) {
      markAsRead();
    }

    if (!entry) {
      const { author, permlink } = this.props;
      entry = await getContent(author, permlink);
    }

    if (entry) {
      const { category, author, permlink } = entry;
      actions.setVisitingEntry(entry);
      history.push(makePath(category, author, permlink, toReplies));
    }
  };

  render() {
    const { children } = this.props;

    return React.cloneElement(children, {
      onClick: this.goEntry
    });
  }
}

EntryLink.defaultProps = {
  entry: null,
  toReplies: false,
  markAsRead: null
};

EntryLink.propTypes = {
  children: PropTypes.element.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  entry: PropTypes.shape({
    category: PropTypes.string.isRequired,
    author: PropTypes.string.isRequired,
    permlink: PropTypes.string.isRequired
  }),
  author: PropTypes.string.isRequired,
  permlink: PropTypes.string.isRequired,
  toReplies: PropTypes.bool,
  actions: PropTypes.shape({
    setVisitingEntry: PropTypes.func.isRequired
  }).isRequired,
  markAsRead: PropTypes.func
};

export default EntryLink;
