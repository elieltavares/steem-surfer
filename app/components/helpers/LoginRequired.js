import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

import { Modal } from 'antd';
import { FormattedMessage, FormattedHTMLMessage } from 'react-intl';

import Login from '../dialogs/Login';

export const checkLogin = (activeAccount, requiredKeys) => {
  // Steem connect login.
  if (activeAccount && activeAccount.type === 'sc') {
    return true;
  }

  // Traditional login. Check all required keys exists in user keys.
  if (
    activeAccount &&
    activeAccount.type === 's' &&
    requiredKeys.every(e => activeAccount.keys[e])
  ) {
    return true;
  }

  return false;
};

class LoginRequired extends Component {
  constructor(props) {
    super(props);

    this.state = {
      loginModalVisible: false
    };
  }

  showLoginModal = () => {
    this.setState({
      loginModalVisible: true
    });

    setTimeout(() => {
      const { onDialogOpen } = this.props;
      onDialogOpen();
    }, 200);
  };

  onLoginModalCancel = () => {
    this.setState({
      loginModalVisible: false
    });
  };

  onLoginSuccess = () => {
    this.setState({
      loginModalVisible: false
    });

    setTimeout(() => {
      if (!this.check()) {
        this.showLoginModal();
      }
    }, 400);
  };

  afterModalClose = () => {
    const { onDialogClose } = this.props;
    onDialogClose();
  };

  check = () => {
    const { activeAccount, requiredKeys } = this.props;

    return checkLogin(activeAccount, requiredKeys);
  };

  render() {
    const { children } = this.props;

    if (this.check()) {
      return children;
    }

    const { activeAccount, requiredKeys } = this.props;
    const { loginModalVisible } = this.state;

    // Default message
    let loginMsg = <FormattedMessage id="login-required.default" />;

    if (activeAccount) {
      // More specific, key based message
      loginMsg =
        requiredKeys.length > 1 ? (
          <FormattedHTMLMessage
            id="login-required.keys-required"
            values={{ keys: requiredKeys }}
          />
        ) : (
          <FormattedHTMLMessage
            id="login-required.key-required"
            values={{ key: requiredKeys }}
          />
        );
    }

    // Clone child element rewriting onclick event
    const newChild = React.cloneElement(children, {
      onClick: () => {
        this.showLoginModal();
      }
    });

    return (
      <Fragment>
        {newChild}

        <Modal
          visible={loginModalVisible}
          onCancel={this.onLoginModalCancel}
          footer={false}
          width="500px"
          closable={false}
          destroyOnClose
          centered
          afterClose={this.afterModalClose}
        >
          <Login
            {...this.props}
            onSuccess={this.onLoginSuccess}
            loginMsg={loginMsg}
          />
        </Modal>
      </Fragment>
    );
  }
}

LoginRequired.defaultProps = {
  activeAccount: null,
  onDialogOpen: () => {},
  onDialogClose: () => {},
  requiredKeys: []
};

LoginRequired.propTypes = {
  children: PropTypes.element.isRequired,
  activeAccount: PropTypes.instanceOf(Object),
  requiredKeys: PropTypes.arrayOf(PropTypes.string),
  onDialogOpen: PropTypes.func,
  onDialogClose: PropTypes.func
};

export default LoginRequired;
