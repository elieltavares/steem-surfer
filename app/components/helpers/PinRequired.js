import React, { Component, Fragment } from 'react';
import { Modal } from 'antd';

import PropTypes from 'prop-types';
import PinConfirm from '../dialogs/PinConfirm';

class PinRequired extends Component {
  constructor(props) {
    super(props);

    this.state = {
      show: false
    };
  }

  onPinInvalidated = () => {
    this.setState({ show: false });
  };

  onConfirmPinSuccess = value => {
    const { onSuccess } = this.props;
    this.setState({ show: false }, () => {
      onSuccess(value);
    });
  };

  cancelFn = () => {
    this.setState({ show: false });
  };

  render() {
    const { children } = this.props;
    const { show } = this.state;

    // Clone child element rewriting onclick event
    const newChild = React.cloneElement(children, {
      onClick: () => {
        this.setState({ show: true });
      }
    });

    return (
      <Fragment>
        {show && (
          <Modal
            footer={null}
            closable
            onCancel={this.cancelFn}
            keyboard
            visible
            width="500px"
            centered
            destroyOnClose
          >
            <PinConfirm
              {...this.props}
              onSuccess={this.onConfirmPinSuccess}
              onInvalidate={this.onPinInvalidated}
            />
          </Modal>
        )}
        {newChild}
      </Fragment>
    );
  }
}

PinRequired.defaultProps = {
  onSuccess: () => {}
};

PinRequired.propTypes = {
  children: PropTypes.element.isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  onSuccess: PropTypes.func
};

export default PinRequired;
