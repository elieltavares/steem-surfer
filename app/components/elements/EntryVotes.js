import React, { Component } from 'react';

import PropTypes from 'prop-types';

import { FormattedMessage, FormattedRelative } from 'react-intl';
import { Popover, Modal, Table, Badge } from 'antd';

import EntryVotesContent from './EntryVotesContent';
import FormattedCurrency from './FormattedCurrency';

import parseToken from '../../utils/parse-token';
import parseDate from '../../utils/parse-date';
import authorReputation from '../../utils/author-reputation';

import AccountLink from '../helpers/AccountLink';
import LinearProgress from '../common/LinearProgress';

import { getActiveVotes, getAccounts } from '../../backend/steem-client';

export const prepareVotes = entry => {
  const totalPayout =
    parseToken(entry.pending_payout_value) +
    parseToken(entry.total_payout_value) +
    parseToken(entry.curator_payout_value);

  const voteRshares = entry.active_votes.reduce(
    (a, b) => a + parseFloat(b.rshares),
    0
  );
  const ratio = totalPayout / voteRshares;

  return entry.active_votes
    .map(a => {
      const rew = a.rshares * ratio;

      return Object.assign({}, a, {
        reputation: authorReputation(a.reputation),
        reward: Number.isNaN(rew) ? 0 : rew,
        time: parseDate(a.time),
        percent: a.percent / 100
      });
    })
    .sort((a, b) => {
      const keyA = a.reward;
      const keyB = b.reward;

      if (keyA > keyB) return -1;
      if (keyA < keyB) return 1;
      return 0;
    });
};

class EntryVotes extends Component {
  constructor(props) {
    super(props);

    this.state = {
      modalVisible: false,
      enabled: false,
      hovered: false,
      realVotes: undefined,
      votesRefreshing: false
    };
  }

  enable = () => {
    this.setState({
      enabled: true,
      hovered: true
    });
  };

  mouseLeave = () => {
    this.setState({
      hovered: false
    });
  };

  showModal = () => {
    this.setState({
      modalVisible: true
    });

    this.setState({ votesRefreshing: true });
    const { entry } = this.props;
    return getActiveVotes(entry.author, entry.permlink)
      .then(votes => {
        const ac = votes.map(a => a.voter);
        return getAccounts(ac).then(accounts =>
          votes.map(v => {
            const account = accounts.find(a => a.name === v.voter);
            if (account) {
              return Object.assign({}, v, { reputation: account.reputation });
            }
            return v;
          })
        );
      })
      .then(votes => {
        const tempEntry = Object.assign({}, entry, { active_votes: votes });
        this.setState({ realVotes: prepareVotes(tempEntry) });
        return votes;
      })
      .catch(() => {})
      .finally(() => {
        this.setState({ votesRefreshing: false });
      });
  };

  handleModalCancel = () => {
    this.setState({
      modalVisible: false
    });
  };

  afterModalClosed = () => {
    this.setState({
      enabled: false
    });
  };

  render() {
    const { entry, children } = this.props;
    if (entry.active_votes.length === 0) {
      // When content has no votes
      return children;
    }

    const {
      modalVisible,
      enabled,
      realVotes,
      votesRefreshing,
      hovered
    } = this.state;

    let popoverProps = {};
    let votes = [];

    if (enabled) {
      votes = realVotes || prepareVotes(entry);

      if (!modalVisible) {
        const popoverContent = (
          <EntryVotesContent {...this.props} votes={votes} />
        );
        popoverProps = { content: popoverContent };
      } else {
        popoverProps.visible = false;
      }
    }

    const clonedChildren = React.cloneElement(children, {
      onClick: this.showModal,
      onMouseEnter: this.enable,
      onMouseLeave: this.mouseLeave
    });

    const modalTableColumns = [
      {
        title: <FormattedMessage id="entry-votes.author" />,
        dataIndex: 'voter',
        key: 'voter',
        width: 220,
        sorter: (a, b) => a.reputation - b.reputation,
        render: (text, record) => (
          <span>
            <AccountLink {...this.props} username={text}>
              <span style={{ cursor: 'pointer', fontWeight: 'bold' }}>
                {text}
              </span>
            </AccountLink>{' '}
            <Badge
              count={record.reputation}
              style={{
                backgroundColor: '#fff',
                color: '#999',
                boxShadow: '0 0 0 1px #d9d9d9 inset'
              }}
            />
          </span>
        )
      },
      {
        title: <FormattedMessage id="entry-votes.reward" />,
        dataIndex: 'reward',
        key: 'reward',
        align: 'center',
        width: 150,
        defaultSortOrder: 'descend',
        sorter: (a, b) => a.reward - b.reward,
        render: text => (
          <FormattedCurrency {...this.props} value={text} fixAt={3} />
        )
      },
      {
        title: <FormattedMessage id="entry-votes.percent" />,
        dataIndex: 'percent',
        key: 'percent',
        align: 'center',
        width: 150,
        sorter: (a, b) => a.percent - b.percent,
        render: text => `${text.toFixed(1)}%`
      },
      {
        title: <FormattedMessage id="entry-votes.time" />,
        dataIndex: 'time',
        key: 'time',
        align: 'center',
        sorter: (a, b) => a.time - b.time,
        render: text => <FormattedRelative value={text} />
      }
    ];

    return (
      <Popover {...popoverProps} visible={hovered && !modalVisible}>
        {clonedChildren}
        <Modal
          visible={modalVisible}
          onCancel={this.handleModalCancel}
          afterClose={this.afterModalClosed}
          footer={false}
          width="80%"
          centered
          title={
            <FormattedMessage
              id="entry-votes.modal-title"
              values={{ n: votes.length }}
            />
          }
        >
          {votesRefreshing && <LinearProgress />}
          <Table
            dataSource={votes}
            columns={modalTableColumns}
            scroll={{ y: 310 }}
            rowKey="voter"
          />
        </Modal>
      </Popover>
    );
  }
}

EntryVotes.propTypes = {
  entry: PropTypes.shape({
    pending_payout_value: PropTypes.string.isRequired,
    total_payout_value: PropTypes.string.isRequired,
    curator_payout_value: PropTypes.string.isRequired,
    active_votes: PropTypes.array.isRequired
  }).isRequired,
  children: PropTypes.element.isRequired
};

export default EntryVotes;
