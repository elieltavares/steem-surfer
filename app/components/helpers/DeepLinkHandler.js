import React from 'react';
import PropTypes from 'prop-types';

import filters from '../../constants/filters.json';

import { makePath as makePathTag } from './TagLink';
import { makePath as makePathAccount } from './AccountLink';
import { makePath as makePathEntry } from './EntryLink';

export const deepLink2Obj = url => {
  let urlPart = url.split('://')[1];

  // remove last char if /
  if (urlPart.endsWith('/')) {
    urlPart = urlPart.substring(0, urlPart.length - 1);
  }

  const parts = urlPart.split('/');

  // witnesses
  if (parts.length === 1 && parts[0] === 'witnesses') {
    return { type: 'witnesses' };
  }

  // sps
  if (parts.length === 1 && parts[0] === 'sps') {
    return { type: 'sps' };
  }

  // filter
  if (parts.length === 1 && filters.includes(parts[0])) {
    return { type: 'filter', filter: parts[0] };
  }

  // filter with tag
  if (parts.length === 2 && filters.includes(parts[0])) {
    return { type: 'filter-tag', filter: parts[0], tag: parts[1] };
  }

  // account
  if (parts.length === 1 && parts[0].startsWith('@')) {
    return { type: 'account', account: parts[0].replace('@', '') };
  }

  // post without tag
  if (parts.length === 2 && parts[0].startsWith('@')) {
    return {
      type: 'post',
      cat: 'esteem',
      author: parts[0].replace('@', ''),
      permlink: parts[1]
    };
  }

  // post with tag
  if (parts.length === 3 && parts[1].startsWith('@')) {
    return {
      type: 'post',
      cat: parts[0],
      author: parts[1].replace('@', ''),
      permlink: parts[2]
    };
  }
};

class DeepLinkHandler extends React.Component {
  componentDidMount() {
    window.addEventListener('deep-link', this.handler);
  }

  componentWillUnmount() {
    window.removeEventListener('deep-link', this.handler);
  }

  handler = e => {
    const { history } = this.props;
    const { url } = e.detail;
    const obj = deepLink2Obj(url);

    let path;

    switch (obj.type) {
      case 'filter':
        path = `/${obj.filter}`;
        break;
      case 'filter-tag':
        path = makePathTag(obj.filter, obj.tag);
        break;
      case 'account':
        path = makePathAccount(obj.account);
        break;
      case 'post':
        path = makePathEntry(obj.cat, obj.author, obj.permlink);
        break;
      case 'sps':
        path = `/sps`;
        break;
      case 'witnesses':
        path = `/witnesses`;
        break;
      default:
        path = null;
        break;
    }

    if (path) {
      history.push(path);
    }
  };

  render() {
    return null;
  }
}

DeepLinkHandler.propTypes = {
  history: PropTypes.shape({}).isRequired
};

export default DeepLinkHandler;
