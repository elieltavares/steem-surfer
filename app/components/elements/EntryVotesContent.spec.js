/* eslint-disable */

import React from 'react';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { getTestData } from '../../../test/data';
import { prepareVotes } from './EntryVotes';
import EntryVotesContent from './EntryVotesContent';
import { mountWithIntl } from '../../helpers/intl-test';

Enzyme.configure({ adapter: new Adapter() });

describe('EntryVoters', () => {
  it('(1) Should render list prepared top 10 votes. Should show more text at bottom.', () => {
    const testData = getTestData(
      'esteemapp',
      'meet-our-cmo-and-product-lead-fil-dunsky-db1db689675c5est'
    );
    const votes = prepareVotes(testData, 1);

    const props = {
      votes: votes,
      global: { currencyRate: 1, currencySymbol: '$' }
    };
    const wrapper = mountWithIntl(<EntryVotesContent {...props} />);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('(2) Should not show more text at bottom. ', () => {
    const testData = getTestData('talhasch', 'one-more-test');
    const votes = prepareVotes(testData, 1);

    const props = {
      votes: votes,
      global: { currencyRate: 1, currencySymbol: '$' }
    };
    const wrapper = mountWithIntl(<EntryVotesContent {...props} />);
    expect(wrapper.html()).toMatchSnapshot();
  });
});
