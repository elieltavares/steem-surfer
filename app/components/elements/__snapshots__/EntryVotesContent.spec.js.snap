// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryVoters (1) Should render list prepared top 10 votes. Should show more text at bottom. 1`] = `"<div class=\\"votes-popover-content\\"><p><span class=\\"voter\\">esteemapp</span><span class=\\"percent\\">100.0%</span><span class=\\"reward\\">$ 22.095</span></p><p><span class=\\"voter\\">good-karma</span><span class=\\"percent\\">100.0%</span><span class=\\"reward\\">$ 9.171</span></p><p><span class=\\"voter\\">esteem.app</span><span class=\\"percent\\">100.0%</span><span class=\\"reward\\">$ 2.624</span></p><p><span class=\\"voter\\">alexis555</span><span class=\\"percent\\">25.0%</span><span class=\\"reward\\">$ 2.041</span></p><p><span class=\\"voter\\">fabien</span><span class=\\"percent\\">100.0%</span><span class=\\"reward\\">$ 1.826</span></p><p><span class=\\"voter\\">cryptoctopus</span><span class=\\"percent\\">19.9%</span><span class=\\"reward\\">$ 0.899</span></p><p><span class=\\"voter\\">dunsky</span><span class=\\"percent\\">100.0%</span><span class=\\"reward\\">$ 0.529</span></p><p><span class=\\"voter\\">khanza.aulia</span><span class=\\"percent\\">100.0%</span><span class=\\"reward\\">$ 0.335</span></p><p><span class=\\"voter\\">boxcarblue</span><span class=\\"percent\\">100.0%</span><span class=\\"reward\\">$ 0.320</span></p><p><span class=\\"voter\\">yahialababidi</span><span class=\\"percent\\">100.0%</span><span class=\\"reward\\">$ 0.221</span></p><p class=\\"more\\"><span>780 more</span></p></div>"`;

exports[`EntryVoters (2) Should not show more text at bottom.  1`] = `"<div class=\\"votes-popover-content\\"><p><span class=\\"voter\\">burhan786</span><span class=\\"percent\\">100.0%</span><span class=\\"reward\\">$ 0.000</span></p><p><span class=\\"voter\\">goidt</span><span class=\\"percent\\">-100.0%</span><span class=\\"reward\\">$ 0.000</span></p></div>"`;
